core/peerset/src/peersstate.rs

```diff
@@ -0,0 +1,479 @@
+/// has a reputation and is either connected to us or not.
+///
+#[derive(Debug, Clone)]
+pub struct PeersState {
+	/// List of nodes that we know about.
+	///
+	/// > **Note**: This list should really be ordered by decreasing reputation, so that we can
+	/// 			easily select the best node to connect to. As a first draft, however, we don't
+	/// 			sort, to make the logic easier.
+	nodes: HashMap<PeerId, Node>,
+
+	/// Maximum allowed number of non-reserved nodes for which the `ConnectionState` is `In`.
+	max_in: u32,
+
+	/// Maximum allowed number of non-reserved nodes for which the `ConnectionState` is `Out`.
+	max_out: u32,
+}
+
+/// State of a single node that we know about.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+struct Node {
+	/// Whether we are connected to this node.
+	connection_state: ConnectionState,
+
+	/// If true, this node is reserved and should always be connected to.
+	reserved: bool,
+
+	/// Reputation value of the node, between `i32::min_value` (we hate that node) and
+	/// `i32::max_value` (we love that node).
+	reputation: i32,
+}
+
+/// Whether we are connected to a node.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+enum ConnectionState {
+	/// We are connected through an ingoing connection.
+	In,
+	/// We are connected through an outgoing connection.
+	Out,
+	/// We are not connected to this node.
+	NotConnected,
+}
+
+impl ConnectionState {
+	/// Returns `true` for `In` and `Out`.
+	fn is_connected(self) -> bool {
+		match self {
+			ConnectionState::In => true,
+			ConnectionState::Out => true,
+			ConnectionState::NotConnected => false,
+		}
+	}
+}
+
+impl PeersState {
+	/// Builds a new empty `PeersState`.
+	pub fn new(in_peers: u32, out_peers: u32) -> Self {
+		PeersState {
+			nodes: HashMap::new(),
+			max_in: in_peers,
+			max_out: out_peers,
+		}
+	}
+
+	/// Adds `value` to the reputation of all the nodes we are connected to.
+	///
+	/// In case of overflow, the value of capped.
+	pub fn connected_reputation_increase(&mut self, value: i32) {
+		for (_, peer_state) in self.nodes.iter_mut() {
+			if peer_state.connection_state.is_connected() {
+				peer_state.reputation = peer_state.reputation.saturating_add(value);
+			}
+		}
+	}
+
+	/// Returns an object that grants access to the state of a peer.
+	pub fn peer<'a>(&'a mut self, peer_id: &'a PeerId) -> Peer<'a> {
+		if let Some(node) = self.nodes.get(peer_id) {
+			if node.connection_state.is_connected() {
+				Peer::Connected(ConnectedPeer {
+					parent: self,
+					peer_id: Cow::Borrowed(peer_id),
+				})
+			} else {
+				Peer::NotConnected(NotConnectedPeer {
+					parent: self,
+					peer_id: Cow::Borrowed(peer_id),
+				})
+			}
+		} else {
+			Peer::Unknown(UnknownPeer {
+				parent: self,
+				peer_id: Cow::Borrowed(peer_id),
+			})
+		}
+	}
+
+	/// Returns the list of all the peers we know of.
+	// Note: this method could theoretically return a `Peer`, but implementing that
+	// isn't simple.
```

Comment:
No, because `Peer` mutably borrows the `PeersState` and we return multiple peers at once.
We can change `Peer` to only borrow the specific entry, but this has some other implications, or we can use some sort of `Rc<RefCell<&mut PeersState>>` instead of just `&mut PeersState`.